### ISSUE TITLE
No way to reduce service probe timeouts at runtime

### ISSUE THREAD

**Issue by djcater** [nmap/nmap#1907]
Unless I'm missing something, there does not seem to be a command-line parameter to reduce the time that a service probe waits for a response.

This is useful when you want to sacrifice a bit of thoroughness for a significant reduction in scan time.

An example is when up against an IPS / tarpit which responds with `SYN-ACK` to all ports for a host / hosts, but black-holes every actual request to that port without responding (beyond `ACK`-ing the TCP packets).

If you know that there is such a device, but that there are genuine open ports within the noise, then it would be useful to reduce the service probe timeout to stop Nmap spending 5 seconds on every single service probe on every single port (but still getting useful data back for a port which is actually open with a legitimate running service).

Here's an example to demonstrate the issue using Nmap 7.80SVN from the current trunk.

Use Ncat to represent an open port that just discards any incoming requests (like a discard service):

```
sudo ncat -4 --verbose --listen --keep-open --recv-only 127.0.0.1 80 > /dev/null
```

Then do a version scan against this open port with Nmap:
```
time sudo nmap -4 -vvv -dd -sS -p80 -sV 127.0.0.1
```

With `-dd` this will show each probe and its timeout, showing a lot of 5000ms timeouts:
```
Initiating Service scan at 23:48
Scanning 1 service on localhost (127.0.0.1)
NSOCK INFO [0.3720s] nsock_iod_new2(): nsock_iod_new (IOD #1)
Starting probes against new service: 127.0.0.1:80 (tcp)
NSOCK INFO [0.3720s] nsock_connect_tcp(): TCP connection requested to 127.0.0.1:80 (IOD #1) EID 8
NSOCK INFO [0.3720s] nsock_trace_handler_callback(): Callback: CONNECT SUCCESS for EID 8 [127.0.0.1:80]
Service scan sending probe NULL to 127.0.0.1:80 (tcp)
NSOCK INFO [0.3720s] nsock_read(): Read request from IOD #1 [127.0.0.1:80] (timeout: 6000ms) EID 18
NSOCK INFO [6.3780s] nsock_trace_handler_callback(): Callback: READ TIMEOUT for EID 18 [127.0.0.1:80]
Service scan sending probe GetRequest to 127.0.0.1:80 (tcp)
NSOCK INFO [6.3780s] nsock_write(): Write request for 18 bytes to IOD #1 EID 27 [127.0.0.1:80]
NSOCK INFO [6.3780s] nsock_read(): Read request from IOD #1 [127.0.0.1:80] (timeout: 5000ms) EID 34
NSOCK INFO [6.3780s] nsock_trace_handler_callback(): Callback: WRITE SUCCESS for EID 27 [127.0.0.1:80]
NSOCK INFO [11.3840s] nsock_trace_handler_callback(): Callback: READ TIMEOUT for EID 34 [127.0.0.1:80]
NSOCK INFO [11.3840s] nsock_iod_delete(): nsock_iod_delete (IOD #1)
NSOCK INFO [11.3840s] nsock_iod_new2(): nsock_iod_new (IOD #2)
NSOCK INFO [11.3840s] nsock_connect_tcp(): TCP connection requested to 127.0.0.1:80 (IOD #2) EID 40
NSOCK INFO [11.3840s] nsock_trace_handler_callback(): Callback: CONNECT SUCCESS for EID 40 [127.0.0.1:80]
Service scan sending probe HTTPOptions to 127.0.0.1:80 (tcp)
NSOCK INFO [11.3840s] nsock_write(): Write request for 22 bytes to IOD #2 EID 51 [127.0.0.1:80]
NSOCK INFO [11.3840s] nsock_read(): Read request from IOD #2 [127.0.0.1:80] (timeout: 5000ms) EID 58
NSOCK INFO [11.3840s] nsock_trace_handler_callback(): Callback: WRITE SUCCESS for EID 51 [127.0.0.1:80]
NSOCK INFO [16.3900s] nsock_trace_handler_callback(): Callback: READ TIMEOUT for EID 58 [127.0.0.1:80]
NSOCK INFO [16.3900s] nsock_iod_delete(): nsock_iod_delete (IOD #2)
...
```
Using `-T5` or `--max-rtt-timeout` doesn't affect the probe timeouts here.

I would say 5000ms is a very long time for a lot of the service probes. Perhaps the default could be reduced and save a lot of time without much loss, but for now I am just proposing a command-line option to cap the timeout. Something like `--max-service-probe-timeout=x`.

I've hacked around this in the source code by changing `DEFAULT_SERVICEWAITMS` to 600 instead of 5000 in `service_scan.h` and then ignoring any `totalwaitms` values from the `nmap-service-probes` file in `service_scan.cc`.

This didn't cover NSE version scripts, so I then hardcoded `request_timeout` to 600 in `nselib/comm.lua`.

This took the above scan time down from **3 minutes 23 seconds**, to **32 seconds**, an 86% reduction in scan time, while still allowing 600ms for a response to every service probe.

Running this modified Nmap against scanme.nmap.org with `-p22,80,443,9929,31337 -sV` didn't result in any reduction in services detected, but the scan time reduced from 8.6 seconds to 3.4 seconds (60% reduction in scan time).

Please let me know if there's anything I can do to further debug this. Thank you.

**TL;DR:** Adding `--max-service-probe-timeout=x` to Nmap would allow capping the time a service probe waits for a response, which in some situations will significantly decrease the time taken to scan without significantly impacting the results.

**Comment by fyodor** [comment]
Thanks for your report and investigation!   To some degree, it is intentional/desirability that changing the version detection timeout is a little more work than simply specifying a command flag or -T value.  Consistency in probe timing is important for consistent results, especially is the user submits the fingerprint for integration into our official database.  We want to that the "standard" procedure is used because failing to give sufficient time for a response can change the signature.

We are very consdervative with the 6s timeout because sometimes targets take longer to respond for reasons such as performing reverse-DNS resolution on the client or because they themselves need to timeout before sending an error response (which could be useful for service detection).

Remember that many services and hosts can undergo version detection at once.  So the preferred approach, assuming you have a lot of hosts to scan, would be to use a larger hostgroup size and also --max-parallelism so that you have dozens or hundreds of 6 second waits going on at once.  The huge time decreases (as a percentage) are more for single-host scans than large groups.

So we aren't planning to add an official feature for this now, but it is good that you have documented some techniques for people who know what they're doing and want to do it.  We would ask that anyone who does such a thing not submit their service fingerprints to our system.  Also note that it is fairly easy to change the totalwaitms in nmap-service-probes file if you are sure you want a lower timeout in general.  Nmap also has well documented ways for you to specify that probes file at runtime so it only happens when you want it to.

Anyway, I am closing this because it's not something we are currently planning to add.  But people can still make comments and we can always reconsider and re-open if we change our mind.  It wouldn't be the first time :).  Thanks again!

**Comment by djcater** [comment]
Thanks for the reply. This was with a large `min-hostgroup`, but I agree that a larger `min-parallelism` is probably the better solution to this particular problem.

**Comment by DaddyBigFish** [comment]
this article is what you need, but until nmap include a simple flag to control this value, then you must compile your own nmap 

https://joshua.hu/nmap-speedup-service-scanning-16x
